fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/recordvalidation/RecordValidationTest.java

```diff
@@ -0,0 +1,408 @@
+            // corrupt the two records
+            byte[] key = ValidationTestUtils.getSplitKey(store, result.get(0).getPrimaryKey(), -1);
+            final byte[] value = new byte[] {1, 2, 3, 4, 5};
+            store.ensureContextActive().set(key, value);
+
+            key = ValidationTestUtils.getSplitKey(store, result.get(1).getPrimaryKey(), -1);
+            store.ensureContextActive().set(key, value);
+
+            commit(context);
+        }
+
+        // Validate by primary key
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore store = openSimpleRecordStore(context, hook, formatVersion);
+            RecordValidator valueValidator = new RecordValueValidator(store);
+            // Currently, we don't handle corrupt versions.
+            Exception ex = assertThrows(ExecutionException.class, () -> valueValidator.validateRecordAsync(result.get(0).getPrimaryKey()).get());
+            assertTrue(ex.getCause() instanceof UnknownValidationException);
+            ex = assertThrows(ExecutionException.class, () -> valueValidator.validateRecordAsync(result.get(1).getPrimaryKey()).get());
+            assertTrue(ex.getCause() instanceof UnknownValidationException);
+
+            commit(context);
+        }
+    }
+
+    private static void validateRecordVersion(final FDBRecordStore store, final Tuple primaryKey, final @Nonnull String expectedVersionValidationCode) throws InterruptedException, ExecutionException {
+        RecordValidator versionValidator = new RecordVersionValidator(store);
+        final RecordValidationResult versionValidatorResult = versionValidator.validateRecordAsync(primaryKey).get();
+        if (expectedVersionValidationCode.equals(RecordValidationResult.CODE_VALID)) {
+            assertTrue(versionValidatorResult.isValid());
+        } else {
+            assertFalse(versionValidatorResult.isValid());
+        }
+        assertEquals(expectedVersionValidationCode, versionValidatorResult.getErrorCode());
+    }
+
+    private static void validateRecordValue(final FDBRecordStore store, final Tuple primaryKey, final @Nonnull String expectedValueValidationCode) throws InterruptedException, ExecutionException {
+        RecordValidator valueValidator = new RecordValueValidator(store);
+        final RecordValidationResult valueValidatorResult = valueValidator.validateRecordAsync(primaryKey).get();
+        if (expectedValueValidationCode.equals(RecordValidationResult.CODE_VALID)) {
+            assertTrue(valueValidatorResult.isValid());
+        } else {
+            assertFalse(valueValidatorResult.isValid());
+        }
+        assertEquals(expectedValueValidationCode, valueValidatorResult.getErrorCode());
+    }
+
+
+    @Nonnull
+    private List<FDBStoredRecord<Message>> saveRecords(final boolean splitLongRecords, final int formatVersion, final RecordMetaDataHook hook) throws Exception {
```

Comment:
The fact that this returns a list, but various tests have are only interacting with one, makes it harder to understand the tests. If this method called two different methods (`saveShortRecord`, and `saveLongRecord`), the methods that only interact with one could call the appropriate method, and be a lot clearer. The others could treat the result as an opaque list.